docs/pages/blog/react-19-update.md

```diff
@@ -0,0 +1,142 @@
+//   1x during state initialization
+// + 1x when sortedRowsSet is fired
+const expectedCallCount = reactMajor >= 19 ? 2 : 4;
+```
+
+### Performance issue
+
+In React 19, you can access `ref` as a prop for function components and `forwardRef` is not longer needed.
+
+This created an [issue](https://github.com/mui/mui-x/issues/15770) for us, which was spotted by our community member.
+
+Because `ref` is now also a prop, spreading props after the ref prop could potentially override the ref.
+
+The existance of `ref` prop on a `ForwardRef` component, even if `undefined`, makes the component props referentially unstable, breaking the downstream memoizations.
+
+To address this, a `forwardRef` shim was added that enforces correct prop ordering at the type level.
+
+```tsx
+// Compatibility shim that ensures stable props object for forwardRef components
+// Fixes https://github.com/facebook/react/issues/31613
+// We ensure that the ref is always present in the props object (even if that's not the case for older versions of React) to avoid the footgun of spreading props over the ref in the newer versions of React.
+export const forwardRef = <T, P = {}>(
+  render: React.ForwardRefRenderFunction<T, P & { ref: React.Ref<T> }>,
+) => {
+  if (reactMajor >= 19) {
+    const Component = (props: any) => render(props, props.ref ?? null);
+    Component.displayName = render.displayName ?? render.name;
+    return Component as React.ForwardRefExoticComponent<P>;
+  }
+  return React.forwardRef(
+    render as React.ForwardRefRenderFunction<T, React.PropsWithoutRef<P>>,
+  );
+};
+```
+
+The shim provides two key benefits:
+
+1. Type safety - TypeScript will warn if props are spread incorrectly
+2. Forward compatibility - Components using the shim will work correctly in all supported React versions
+
+The code update was:
+
+```tsx
+// Before
+const GridRoot = forwardRef((props, ref) => {
+  const state = useGridState();
+  return <div ref={ref} {...props} {...state} />;
+});
+
+// After
+const GridRoot = forwardRefShim((props, ref) => {
+  const state = useGridState();
+  return <div {...props} {...state} ref={ref} />;
+});
+```
+
+## Phase 2: Moving to React 19
+
+After ensuring compatibility, we started working on migrating the codebase to React 19. This involved:
+
+1. Updating all package dependencies to their React 19 compatible versions
+2. Migrating test utilities to work with React 19
+3. Ensuring all components work with the new React 19 features
+4. Updating CI to run tests with React 18
+5. Type references with `RefObject` for React 19 and with `MutableRefObject` for earlier versions.
+
+Biggest change in this phase was arround the `useRef()` hook [update](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#useref-requires-argument) and the fact that `apiRef` in the DataGrid component had to be updated from `MutableRefObject` to `RefObject` for React 19 only, in order to avoid type errors for users (and their CI) that did not migrate yet.
```

Comment:
```suggestion
The biggest change in this phase was around the `useRef()` hook [update](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#useref-requires-argument) and the fact that `apiRef` in the DataGrid component had to be updated from `MutableRefObject` to `RefObject` for React 19 only, in order to avoid type errors for users (and their CI) that did not migrate yet.
```